File: src/content/docs/topics/cs/coding-problems/tries/212-word-search-ii-approach3.ts
function assert(condition: boolean, msg: string = ''): void {
    if (!condition) throw new Error(msg || 'Assertion failed');
}

class TrieNode {
    children: Map<string, TrieNode> = new Map();
    word: string | null = null;
}

function findWords(board: string[][], words: string[]): string[] {
    const root = new TrieNode();
    for (const w of words) {
        let node = root;
        for (const ch of w) {
            if (!node.children.has(ch)) node.children.set(ch, new TrieNode());
            node = node.children.get(ch)!;
        }
        node.word = w;
    }

    const rows = board.length, cols = board[0].length;
    const found: string[] = [];

    function dfs(r: number, c: number, node: TrieNode): void {
        if (r < 0 || r >= rows || c < 0 || c >= cols) return;
        const ch = board[r][c];
        if (ch === '#' || !node.children.has(ch)) return;
        const nextNode = node.children.get(ch)!;
        if (nextNode.word !== null) { found.push(nextNode.word); nextNode.word = null; }
        board[r][c] = '#';
        dfs(r + 1, c, nextNode); dfs(r - 1, c, nextNode);
        dfs(r, c + 1, nextNode); dfs(r, c - 1, nextNode);
        board[r][c] = ch;
        // Dead-branch pruning
        if (nextNode.children.size === 0 && nextNode.word === null) {
            node.children.delete(ch);
        }
    }

    for (let r = 0; r < rows; r++)
        for (let c = 0; c < cols; c++)
            dfs(r, c, root);
    return found;
}

const board1 = [["o","a","a","n"],["e","t","a","e"],["i","h","k","r"],["i","f","l","v"]];
const r1 = new Set(findWords(board1, ["oath","pea","eat","rain"]));
assert(r1.has("oath") && r1.has("eat") && r1.size === 2);
assert(JSON.stringify(findWords([["a"]], ["a"])) === JSON.stringify(["a"]));
assert(JSON.stringify(findWords([["a"]], ["b"])) === JSON.stringify([]));
const board2 = [["a","b"],["c","d"]];
const r2 = new Set(findWords(board2, ["ab","cd","abdc"]));
assert(r2.has("ab") && r2.has("cd") && r2.has("abdc") && r2.size === 3);
console.log("all tests pass");

File: src/content/docs/topics/cs/coding-problems/tries/212-word-search-ii-approach3.go
package main

import (
	"fmt"
	"sort"
)

// Approach 3: Trie + board DFS + dead-branch pruning

type TrieNode struct {
	Children [26]*TrieNode
	Word     string // non-empty at terminal nodes
}

func findWords(board [][]byte, words []string) []string {
	root := &TrieNode{}
	for _, w := range words {
		node := root
		for _, ch := range w {
			i := ch - 'a'
			if node.Children[i] == nil {
				node.Children[i] = &TrieNode{}
			}
			node = node.Children[i]
		}
		node.Word = w
	}

	rows, cols := len(board), len(board[0])
	var found []string

	var dfs func(r, c int, node *TrieNode)
	dfs = func(r, c int, node *TrieNode) {
		if r < 0 || r >= rows || c < 0 || c >= cols {
			return
		}
		ch := board[r][c]
		if ch == '#' {
			return
		}
		i := ch - 'a'
		if node.Children[i] == nil {
			return
		}
		next := node.Children[i]
		if next.Word != "" {
			found = append(found, next.Word)
			next.Word = ""
		}
		board[r][c] = '#'
		dfs(r+1, c, next)
		dfs(r-1, c, next)
		dfs(r, c+1, next)
		dfs(r, c-1, next)
		board[r][c] = ch
		// Dead-branch pruning: remove leaf with no children and no word
		if next.Word == "" && allNil(next.Children[:]) {
			node.Children[i] = nil
		}
	}

	for r := 0; r < rows; r++ {
		for c := 0; c < cols; c++ {
			dfs(r, c, root)
		}
	}
	return found
}

func allNil(children []*TrieNode) bool {
	for _, c := range children {
		if c != nil {
			return false
		}
	}
	return true
}

func assert(condition bool, msgs ...string) {
	if !condition {
		msg := "assertion failed"
		if len(msgs) > 0 {
			msg = msgs[0]
		}
		panic(msg)
	}
}

func setsEqual(a, b []string) bool {
	if len(a) != len(b) {
		return false
	}
	sort.Strings(a)
	sort.Strings(b)
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

func strBoard(rows [][]string) [][]byte {
	b := make([][]byte, len(rows))
	for i, row := range rows {
		b[i] = make([]byte, len(row))
		for j, s := range row {
			b[i][j] = s[0]
		}
	}
	return b
}

func runTests() {
	board1 := strBoard([][]string{{"o", "a", "a", "n"}, {"e", "t", "a", "e"}, {"i", "h", "k", "r"}, {"i", "f", "l", "v"}})
	assert(setsEqual(findWords(board1, []string{"oath", "pea", "eat", "rain"}), []string{"eat", "oath"}))
	assert(setsEqual(findWords([][]byte{{'a'}}, []string{"a"}), []string{"a"}))
	assert(setsEqual(findWords([][]byte{{'a'}}, []string{"b"}), []string{}))
	board2 := strBoard([][]string{{"a", "b"}, {"c", "d"}})
	assert(setsEqual(findWords(board2, []string{"ab", "cd", "abdc"}), []string{"ab", "abdc", "cd"}))
	fmt.Println("all tests pass")
}

func main() { runTests() }
